### PR TITLE
Adding full logic related to user education

### PR DESCRIPTION
**Pull request by BelehovEgor**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/322

### CODE REVIEW

**Review by TIove — CHANGES_REQUESTED** [review]

File: src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    public class CreateEducationRequestValidator : AbstractValidator<CreateEducationRequest>, ICreateEducationRequestValidator
+    {
+        public CreateEducationRequestValidator(IUserRepository repository)
+        {
+            RuleFor(education => education.UserId)
+                .Must(id => repository.IsExistUser(id));
+
+            RuleFor(education => education.UniversityName)
+                .NotEmpty()
+                .MaximumLength(50).WithMessage("University name is too long");
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs:17]: and how many do you want, if it too little for you?

File: test/UserService.Data.UnitTests/UserRepositoryTests.cs [hunk 2]
```diff
@@ -191,21 +191,22 @@ public void ShouldThrowExeptionWhenUserWasNotFound()
             Assert.Throws<NotFoundException>(() => _repository.EditUser(userId, _userPatch));
         }
 
-        [Test]
-        public void ShouldEditUser()
-        {
-            bool isEdit =_repository.EditUser(_dbUser.Id, _userPatch);
+        //TODO
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply test/UserService.Data.UnitTests/UserRepositoryTests.cs:194]: It is not my test, but it didn't work normal if I test all, it is a failure if I test only it, it is successful

File: src/UserService.Models.Dto/Requests/User/Education/EditEducationRequest.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+﻿using LT.DigitalOffice.UserService.Models.Dto.Enums;
+using System;
+
+namespace LT.DigitalOffice.UserService.Models.Dto.Requests.User.Education
+{
+    public class EditEducationRequest
+    {
+        public string UniversityName { get; set; }
+        public string QualificationName { get; set; }
+        public FormEducation FormEducation { get; set; }
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Models.Dto/Requests/User/Education/EditEducationRequest.cs:10]: where?

File: src/UserService.Business/Commands/User/Education/EditEducationCommand.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            _accessValidator = accessValidator;
+            _httpContextAccessor = httpContextAccessor;
+            _repository = repository;
+            _mapper = mapper;
+            _validator = validator;
+        }
+
+        public OperationResultResponse<bool> Execute(Guid userId, Guid educationId, JsonPatchDocument<EditEducationRequest> request)
+        {
+            if (!(_accessValidator.IsAdmin() ||
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Business/Commands/User/Education/EditEducationCommand.cs:44]: why will it be better?

File: src/UserService.Models.Dto/Models/EducationInfo.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace LT.DigitalOffice.UserService.Models.Dto.Models
+{
+    public class EducationInfo
+    {
+        public Guid Id { get; set; }
+        public string UniversityName { get; set; }
+        public string QualificationName { get; set; }
+        public FormEducation FormEducation { get; set; }
+        public DateTime AdmissiomAt { get; set; }
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Models.Dto/Models/EducationInfo.cs]: fixed

File: src/UserService.Models.Db/DbUserEducation.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    public class DbUserEducation
+    {
+        public const string TableName = "UserEducations";
+
+        public Guid Id { get; set; }
+        public Guid UserId { get; set; }
+        public string UniversityName { get; set; }
+        public string QualificationName { get; set; }
+        public int FormEducation { get; set; }
+        public DateTime AdmissiomAt { get; set; }
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Models.Db/DbUserEducation.cs]: fixed

File: src/UserService.Models.Dto/Requests/User/Education/CreateEducationRequest.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace LT.DigitalOffice.UserService.Models.Dto.Requests.User.Education
+{
+    public class CreateEducationRequest
+    {
+        public Guid UserId { get; set; }
+        public string UniversityName { get; set; }
+        public string QualificationName { get; set; }
+        public FormEducation FormEducation { get; set; }
+        public DateTime AdmissiomAt { get; set; }
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Models.Dto/Requests/User/Education/CreateEducationRequest.cs]: fixed

File: src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs:17]: now 100

File: src/UserService.Validation/User/Education/EditEducationRequestValidator.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                .DependentRules(() =>
+                {
+                    When(x => GetOperationByPath(x, UniversityName) != null, () =>
+                    {
+                        RuleFor(x => x.Operations)
+                            .UniqueOperationWithAllowedOp(UniversityName, "replace");
+
+                        RuleFor(x => (string)GetOperationByPath(x, UniversityName).value)
+                            .NotEmpty()
+                            .MaximumLength(50)
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Validation/User/Education/EditEducationRequestValidator.cs]: now 100

File: src/UserService.Validation/User/Education/EditEducationRequestValidator.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+                    });
+
+                    When(x => GetOperationByPath(x, QualificationName) != null, () =>
+                    {
+                        RuleFor(x => x.Operations)
+                            .UniqueOperationWithAllowedOp(QualificationName, "replace");
+
+                        RuleFor(x => (string)GetOperationByPath(x, QualificationName).value)
+                            .NotEmpty()
+                            .MaximumLength(50)
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Validation/User/Education/EditEducationRequestValidator.cs]: now 100

File: src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+            RuleFor(education => education.QualificationName)
+                .NotEmpty()
+                .MaximumLength(50).WithMessage("Qualification name is too long");
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs:21]: now 100

File: test/UserService.Mappers.UnitTests/Requests/Education/DbUserEducationMapperTests.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            var result = _mapper.Map(_education);
+
+            _dbEducation.Id = result.Id;
+
+            SerializerAssert.AreEqual(_dbEducation, result);
+        }
+
+        [Test]
+        public void ShouldThrowExceptionWhenrequestIdNull()
```
Comment (djino97) [inline test/UserService.Mappers.UnitTests/Requests/Education/DbUserEducationMapperTests.cs:54]: missed uppercase

File: test/UserService.Validation.UnitTests/Education/CreateEducationRequestValidatorTests.cs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            };
+
+            _repositoryMock
+                .Setup(x => x.IsExistUser(It.IsAny<Guid>()))
+                .Returns(false);
+
+            _validator.TestValidate(_request).ShouldHaveValidationErrorFor(x => x.UniversityName);
+            _validator.TestValidate(_request).ShouldHaveValidationErrorFor(x => x.QualificationName);
+            _validator.TestValidate(_request).ShouldHaveValidationErrorFor(x => x.FormEducation);
+            _validator.TestValidate(_request).ShouldHaveValidationErrorFor(x => x.UserId);
```
Comment (djino97) [inline test/UserService.Validation.UnitTests/Education/CreateEducationRequestValidatorTests.cs:64]: What this checking?

File: test/UserService.Validation.UnitTests/Education/EditEducationRequestValidatorTests.cs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+namespace LT.DigitalOffice.UserService.Validation.UnitTests.Education
+{
+    class EditEducationRequestValidatorTests
+    {
+        private EditEducationRequestValidator _validator;
+        private JsonPatchDocument<EditEducationRequest> _request;
+
+        [SetUp]
+        public void SetUp()
```
Comment (djino97) [inline test/UserService.Validation.UnitTests/Education/EditEducationRequestValidatorTests.cs:19]: use OneTimeSetUp

File: src/UserService.Business/Commands/Education/CreateEducationCommand.cs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+                  _accessValidator.HasRights(Rights.AddEditRemoveUsers))
+                  && senderId != request.UserId)
+            {
+                throw new ForbiddenException("Not enough rights.");
+            }
+
+            _validator.ValidateAndThrowCustom(request);
+
+            var dbEducation = _mapper.Map(request);
+
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Business/Commands/Education/CreateEducationCommand.cs:56]: fixed

File: src/UserService.Business/Commands/Education/RemoveEducationCommand.cs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+            var dbUser = _repository.Get(senderId);
+
+            if (!(dbUser.IsAdmin ||
+                  _accessValidator.HasRights(Rights.AddEditRemoveUsers))
+                  && senderId != userId)
+            {
+                throw new ForbiddenException("Not enough rights.");
+            }
+
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Business/Commands/Education/RemoveEducationCommand.cs:42]: please check line 38

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210429102831_AddUserEducation.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+                name: DbUserEducation.TableName,
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(nullable: false),
+                    UserId = table.Column<Guid>(nullable: false),
+                    UniversityName = table.Column<string>(nullable: false),
+                    QualificationName = table.Column<string>(nullable: false),
+                    FormEducation = table.Column<int>(nullable: false),
+                    AdmissionAt = table.Column<DateTime>(nullable: false),
+                    IssueAt = table.Column<DateTime>(nullable: true)
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210429102831_AddUserEducation.cs:24]: yes, it is struct

File: src/UserService.Data/UserRepository.cs [hunk 17]
```diff
@@ -197,5 +203,66 @@ public void DeletePendingUser(Guid userId)
...
+        public bool RemoveEducation(Guid educationId)
+        {
+            DbUserEducation dbUserEducation = _provider.UserEducations.FirstOrDefault(e => e.Id == educationId);
+
+            if (dbUserEducation == null)
+            {
+                throw new NotFoundException($"User education with ID '{educationId}' was not found.");
+            }
+
+            _provider.UserEducations.Remove(dbUserEducation);
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Data/UserRepository.cs:257]: I guess it is not an important entity and we can remove this, do you want to do flag active?

File: src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Validation/User/Education/CreateEducationRequestValidator.cs:17]: fixed

File: src/UserService.Validation/User/Education/EditEducationRequestValidator.cs [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+                .WithMessage("You don't have changes.")
+                .ForEach(y => y
+                    .Must(x => Paths.Any(cur => string.Equals(
+                        cur,
+                        x.path,
+                        StringComparison.OrdinalIgnoreCase)))
+                    .WithMessage(
+                        $"Document contains invalid path. Only such paths are allowed: {Paths.Aggregate((x, y) => x + ", " + y)}")
+                )
+                .DependentRules(() =>
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Validation/User/Education/EditEducationRequestValidator.cs:43]: we cant remove or add properties, only edit them

File: src/UserService/Startup.cs [hunk 19]
```diff
@@ -247,7 +248,7 @@ public void Configure(IApplicationBuilder app, ILoggerFactory loggerFactory)
 
             app.UseRouting();
 
-            app.UseMiddleware<TokenMiddleware>();
+            //app.UseMiddleware<TokenMiddleware>();
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService/Startup.cs:251]: ok)

File: test/UserService.Mappers.UnitTests/Requests/Education/DbUserEducationMapperTests.cs (hunk 11 above)
Comment (BelehovEgor) [reply test/UserService.Mappers.UnitTests/Requests/Education/DbUserEducationMapperTests.cs:54]: fixed

File: test/UserService.Validation.UnitTests/Education/CreateEducationRequestValidatorTests.cs (hunk 12 above)
Comment (BelehovEgor) [reply test/UserService.Validation.UnitTests/Education/CreateEducationRequestValidatorTests.cs:64]: This validator checks all properties and after then return all problems, so this test has errors in every property and I check this

File: test/UserService.Validation.UnitTests/Education/EditEducationRequestValidatorTests.cs (hunk 13 above)
Comment (BelehovEgor) [reply test/UserService.Validation.UnitTests/Education/EditEducationRequestValidatorTests.cs:19]: fixed

File: src/UserService.Business/Commands/Education/CreateEducationCommand.cs [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+            _httpContextAccessor = httpContextAccessor;
+            _mapper = mapper;
+            _repository = repository;
+            _validator = validator;
+        }
+
+        public OperationResultResponse<Guid> Execute(CreateEducationRequest request)
+        {
+            var senderId = _httpContextAccessor.HttpContext.GetUserId();
+
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Business/Commands/Education/CreateEducationCommand.cs:43]: fixed

File: src/UserService.Data/UserRepository.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/UserService.Data/UserRepository.cs]: I wasn't right, fixed

**Review by TIove — APPROVED** [review]

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210429102831_AddUserEducation.cs [hunk 21]
```diff
@@ -21,7 +21,8 @@ private void CreateUserEducationTable(MigrationBuilder migrationBuilder)
                     QualificationName = table.Column<string>(nullable: false),
                     FormEducation = table.Column<int>(nullable: false),
                     AdmissionAt = table.Column<DateTime>(nullable: false),
-                    IssueAt = table.Column<DateTime>(nullable: true)
+                    IssueAt = table.Column<DateTime>(nullable: true),
```
Comment (djino97) [inline src/UserService.Data.Provider.MsSql.Ef/Migrations/20210429102831_AddUserEducation.cs:24]: write DateTime

**Review by djino97 — APPROVED** [review]

**Review by borovHokage — APPROVED** [review]

**Review by KostinP — APPROVED** [review]

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20210429102831_AddUserEducation.cs (hunk 21 above)
Comment (BelehovEgor) [reply src/UserService.Data.Provider.MsSql.Ef/Migrations/20210429102831_AddUserEducation.cs]: fixed

**Pull request merged**